server/src/main/scala/org/http4s/server/middleware/authentication/NonceF.scala

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2014 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.server.middleware.authentication
+
+import cats.effect.Sync
+import cats.effect.concurrent.Ref
+import cats.syntax.all._
+import org.http4s.crypto.unsafe.SecureRandom
+
+import java.math.BigInteger
+import java.util.Date
+
+private[authentication] class NonceF[F[_]](val created: Date, val nc: Ref[F, Int], val data: String)
+
+private[authentication] object NonceF {
+  val random = new SecureRandom()
+
+  private def getRandomData[F[_]](bits: Int)(implicit F: Sync[F]): F[String] =
+    F.delay(new BigInteger(bits, random).toString(16))
+
+  def gen[F[_]: Sync](bits: Int): F[NonceF[F]] =
```

Comment:
This is going to need a `Blocker` and a `ContextShift` in the long run.  This can merge before #6165, but it would be best to get that done before we have a release and need to do a MiMa exemption.

The nice thing is, in Cats-Effect 3, this becomes more transparent in the API.

Reply:
Sort of interesting -- `NonceKeeperF[F].newNonce()` already requires a permit for the singleton semaphore, so I think acquiring an instance of `Blocker` at that time makes sense. I can keep moving it further up if desired, I really don't have a good intuition around where that instance needs to come from.

Reply:
Does `Blocking[F].use` actually acquire a new thread/threadpool? Should that parameter be bubbled all the way out to the caller's code?

Reply:
Cats-Effect 2's `Blocker` does require a thread pool.  One `Blocker` should suffice for most apps, but any app that does blocking needs one.  It's part of the runtime in Cats-Effect 3, and available from the `Sync` constraint, so this will melt away in 0.23.

Engineering tradeoff: if we don't thread it through, this could suck real bad for a small number of users.  If we do thread it through, we're going to all that work for, like, Windows users, and other people who do Weird Shit.  http4s-0.22 is basically EOL.  Personally, I'm only fixing bugs on it.  And this blocking is not new, and nobody has complained.  So if you'd prefer, we could put a fat disclaimer on it and know it will be better in CE3.  But this is "safer".

Reply:
Oh, I overlooked the `Blocker.use`.  Yeah, that creates a new cached thread pool on every invocation.  What we want is one cached blocker for the entire app that gets threaded through.  This is where the API is going to get uglier in CE2.

Reply:
Swapped it out to a single allocated `Blocker` on init, backed by the `global` threadpool, as this is going away anyway.

Reply:
Alright -- threaded it through, as this is a new integration, and the exposure of it will better match the functionality in CE3 for users through the upgrade process.